src/responder/pam/pamsrv_gssapi.c

```diff
@@ -0,0 +1,631 @@
+/*
+    Authors:
+        Pavel Březina <pbrezina@redhat.com>
+
+    Copyright (C) 2020 Red Hat
+
+    This program is free software; you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation; either version 3 of the License, or
+    (at your option) any later version.
+
+    This program is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <errno.h>
+#include <gssapi.h>
+#include <gssapi/gssapi_ext.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <talloc.h>
+#include <ldb.h>
+
+#include "db/sysdb.h"
+#include "responder/common/responder_packet.h"
+#include "responder/common/responder.h"
+#include "responder/common/cache_req/cache_req.h"
+#include "responder/pam/pamsrv.h"
+#include "sss_client/sss_cli.h"
+#include "util/util.h"
+#include "util/sss_utf8.h"
+
+static errno_t read_str(size_t body_len,
+                        uint8_t *body,
+                        size_t *pctr,
+                        const char **_str)
+{
+    size_t i;
+
+    for (i = *pctr; i < body_len && body[i] != 0; i++) {
+        /* counting */
+    }
+
+    if (i >= body_len) {
+        return EINVAL;
+    }
+
+    if (!sss_utf8_check(&body[*pctr], i - *pctr)) {
+        DEBUG(SSSDBG_CRIT_FAILURE, "Body is not UTF-8 string!\n");
+        return EINVAL;
+    }
+
+    *_str = (const char *)&body[*pctr];
+    *pctr = i + 1;
+
+    return EOK;
+}
+
+static bool pam_gssapi_allowed(struct pam_ctx *pam_ctx,
+                               struct sss_domain_info *domain,
+                               const char *service)
+{
+    char **list = pam_ctx->gssapi_services;
+
+    if (domain->gssapi_services != NULL) {
+        list = domain->gssapi_services;
+    }
+
+    return string_in_list(service, list, true);
+}
+
+static char *pam_gssapi_target(TALLOC_CTX *mem_ctx,
+                               struct sss_domain_info *domain)
+{
+    return talloc_asprintf(mem_ctx, "host@%s", domain->hostname);
+}
+
+static const char *pam_gssapi_get_upn(struct cache_req_result *result)
+{
+    const char *upn;
+
+    if (result->count == 0) {
+        return NULL;
+    }
+
+    /* Canonical UPN should be available if the user has kinited through SSSD.
+     * Fallback to general UPN if it is not defined. */
+    upn = ldb_msg_find_attr_as_string(result->msgs[0], SYSDB_CANONICAL_UPN,
```

Comment:
I think only SYSDB_CANONICAL_UPN should be used here and if it is not available just return NULL but do not fail be return an empty name back to pam_sss_gss and let if pick just the active TGT.

I think the actual check if the right principal is used should happen during SSS_GSSAPI_SEC_CTX because this is where the decision is made, during the init step it is just a hint for what we are looking for.

It came to my mind that for a proper check we have to required that the original authentication was done with SSSD (or at least the the user once has logged in with a password). In cases where users are using ssh with GSSAPI authentication and ticket forwarding we might not reliable know the principal. So I think it might be worth to have an option to disable the principal check. We can add more flexibility later e.g. by considering `krb5_user_map` or other options.

bye,
Sumit

Reply:
> I think only SYSDB_CANONICAL_UPN should be used here and if it is not available just return NULL but do not fail be return an empty name back to pam_sss_gss and let if pick just the active TGT.
>
> I think the actual check if the right principal is used should happen during SSS_GSSAPI_SEC_CTX because this is where the decision is made, during the init step it is just a hint for what we are looking for.

How does it differ?

* We return required UPN X@REALMX, related TGT is not found in the ccache, we fail.
* We don't request specific UPN, user is able to get the service ticket with different principal Y@REALMY, we try to match Y@REALMY to X@REALMX and we fail.

Reply:
> > I think only SYSDB_CANONICAL_UPN should be used here and if it is not available just return NULL but do not fail be return an empty name back to pam_sss_gss and let if pick just the active TGT.
> > I think the actual check if the right principal is used should happen during SSS_GSSAPI_SEC_CTX because this is where the decision is made, during the init step it is just a hint for what we are looking for.
> 
> How does it differ?
> 
>     * We return required UPN X@REALMX, related TGT is not found in the ccache, we fail.

Sending the UPN back during init would help GSSAPI to find the right ticket in a cache collection as long as all TGTs are for different realms.

But since this is a 2 step process, init and sec_ctx. In the first step we ask please get a service ticket for service S with principal X. The second step already checks if the service ticket is correct so imo it should not just assume that it was requested with the right principal but should check it as well.

bye,
Sumit

> 
>     * We don't request specific UPN, user is able to get the service ticket with different principal Y@REALMY, we try to match Y@REALMY to X@REALMX and we fail.

Reply:
I see. Done.